docs/book/how-to/handle-data-artifacts/datasets.md

```diff
@@ -0,0 +1,171 @@
+from zenml.materializers import BaseMaterializer
+from zenml.io import fileio
+from zenml.enums import ArtifactType
+import json
+import os
+
+class CSVDatasetMaterializer(BaseMaterializer):
+    ASSOCIATED_TYPES = (CSVDataset,)
+    ASSOCIATED_ARTIFACT_TYPE = ArtifactType.DATA
+
+    def load(self, data_type: Type[CSVDataset]) -> CSVDataset:
+        with fileio.open(os.path.join(self.uri, "metadata.json"), "r") as f:
+            metadata = json.load(f)
+        return CSVDataset(metadata["data_path"])
+
+    def save(self, dataset: CSVDataset) -> None:
+        metadata = {"data_path": dataset.data_path}
+        with fileio.open(os.path.join(self.uri, "metadata.json"), "w") as f:
+            json.dump(metadata, f)
+        if dataset.df is not None:
+            dataset.df.to_csv(dataset.data_path, index=False)
+
+class BigQueryDatasetMaterializer(BaseMaterializer):
+    ASSOCIATED_TYPES = (BigQueryDataset,)
+    ASSOCIATED_ARTIFACT_TYPE = ArtifactType.DATA
+
+    def load(self, data_type: Type[BigQueryDataset]) -> BigQueryDataset:
+        with fileio.open(os.path.join(self.uri, "metadata.json"), "r") as f:
+            metadata = json.load(f)
+        dataset = BigQueryDataset(
+            table_id=metadata["table_id"],
+            project=metadata.get("project"),
+        )
+        dataset.read_data()
+        return dataset
+
+    def save(self, bq_dataset: BigQueryDataset) -> None:
+        metadata = {
+            "table_id": bq_dataset.table_id,
+            "project": bq_dataset.project,
+        }
+        with fileio.open(os.path.join(self.uri, "metadata.json"), "w") as f:
+            json.dump(metadata, f)
+        if bq_dataset.df is not None:
+            bq_dataset.write_data()
+```
+
+## Managing Complexity in Pipelines with Multiple Data Sources
+
+When working with multiple data sources, it's crucial to design flexible pipelines that can handle different scenarios. Here's an example of how to structure a pipeline that works with both CSV and BigQuery datasets:
+
+```python
+from zenml import step, pipeline
+from typing_extensions import Annotated
+
+@step
+def extract_data_local(data_path: str = "data/raw_data.csv") -> CSVDataset:
+    return CSVDataset(data_path)
+
+@step
+def extract_data_remote(table_id: str) -> BigQueryDataset:
+    return BigQueryDataset(table_id)
+
+@step
+def transform_csv(dataset: CSVDataset) -> Annotated[CSVDataset, "transformed_dataset"]:
+    df = dataset.read_data()
+    # Transform data
+    transformed_df = df.copy()  # Apply transformations here
+    return CSVDataset("tmp/transformed_data.csv", df=transformed_df)
+
+@step
+def transform_bq(dataset: BigQueryDataset) -> Annotated[BigQueryDataset, "transformed_dataset"]:
+    df = dataset.read_data()
+    # Transform data
+    transformed_df = df.copy()  # Apply transformations here
+    return BigQueryDataset(table_id="project.dataset.transformed_table", df=transformed_df)
+
+@pipeline
+def etl_pipeline(mode: str = "develop") -> Dataset:
```

Comment:
This does not actually return Dataset, I would just remove the annotation entirely